src/kirin/serialization/base/registry.py

```diff
@@ -0,0 +1,147 @@
+import inspect
+import pkgutil
+import importlib
+from typing import Any, Dict
+from dataclasses import dataclass
+
+from kirin import ir
+from kirin.serialization.base.serializermixin import SerializerMixin
+
+RUNTIME_ENCODE_LOOKUP: Dict[str, Any] = {}
+RUNTIME_DECODE_LOOKUP: Dict[str, Any] = {}
+RUNTIME_NAME2TYPE: Dict[str, type] = {}
+DIALECTS_LOOKUP = {}
+
+
+def autodiscover_serializers() -> None:
+    try:
+        impls_pkg = importlib.import_module("kirin.serialization.base.impls")
+    except Exception:
+        return
+    for finder, module_name, ispkg in pkgutil.walk_packages(
+        impls_pkg.__path__, impls_pkg.__name__ + "."
+    ):
+        try:
+            mod = importlib.import_module(module_name)
+        except Exception:
+            continue
+        for obj in vars(mod).values():
+            if inspect.isclass(obj):
+                if obj is SerializerMixin:
+                    continue
+                if issubclass(obj, SerializerMixin) or (
+                    hasattr(obj, "serialize") and hasattr(obj, "deserialize")
+                ):
+                    RUNTIME_NAME2TYPE[obj.__name__] = obj
+
+
+def register_type(obj_type: type):
+    RUNTIME_NAME2TYPE[obj_type.__name__] = obj_type
+    return obj_type
+
+
+def register_dialect(dialect: ir.Dialect):
+    stmt_map: dict[str, type] = {}
+    for stmt_cls in dialect.stmts:
+        stmt_map[stmt_cls.__name__] = stmt_cls
+        stmt_declared_name = getattr(stmt_cls, "name", None)
+        if stmt_declared_name and stmt_declared_name != stmt_cls.__name__:
+            stmt_map[stmt_declared_name] = stmt_cls
+
+    DIALECTS_LOOKUP[dialect.name] = (dialect, stmt_map)
+
+
+def runtime_register_encode(obj_type):
+    def wrapper(func):
+        import warnings
+
+        warnings.warn(
+            "runtime_register_encode is deprecated and will be removed; "
+            "prefer the new Serializer API.",
+            DeprecationWarning,
+            stacklevel=2,
+        )
+        RUNTIME_ENCODE_LOOKUP[obj_type.__name__] = func
```

Comment:
using the name might be problematic here since an object's name might happen multiple times between different dialects. 

This should at least throw an error if there is a name collision.